File: 골드 5 (3문제)/7569_토마토.py
# 설명 링크 : https://moon-gd.tistory.com/80
import sys
import collections
q_input = sys.stdin.readline  # 빠른 입력을 위해
queue = collections.deque([])

x_size, y_size, z_size = map(int, q_input().split())  # 토마토 상자 크기 확인

# 공간 정보 선언
total_size = x_size * y_size * z_size  # 토마토 상자 전체 칸 개수
empty_size = 0  # 빈 공간 수
colored_size = 0  # 익은 토마토가 있는 칸의 수

# 방문 리스트, 토마토 리스트
visited = [[[False for _ in range(x_size)] for _ in range(y_size)] for _ in range(z_size)]

tomatoes = []

for k in range(z_size):
    temp = []
    for j in range(y_size):
        temp.append([])
    tomatoes.append(temp)

# 방향 정보 리스트
dir_vector = [[0, 0, 1], [0, 0, -1], [0, -1, 0], [1, 0, 0], [0, 1, 0], [-1, 0, 0]]

# 정답 변수
target_day = 0

for k in range(z_size):
    for j in range(y_size):
        x_list = list(map(int, q_input().split()))
        tomatoes[k][j] = x_list
        for i in range(x_size):
            # 익은 토마토가 들어 있는 경우
            if x_list[i] == 1:
                colored_size += 1  # 색칠된 칸 개수 += 1
                queue.append([k, j, i, 0])  # 큐에 넣어주기
                visited[k][j][i] = True  # 방문 배열 체크

            # 빈 칸인 경우
            elif x_list[i] == -1:
                empty_size += 1  # 빈 칸 개수 += 1


while len(queue):
    cur_z, cur_y, cur_x, cur_day = queue.popleft()

    for vector in dir_vector:
        new_x = cur_x + vector[0]
        new_y = cur_y + vector[1]
        new_z = cur_z + vector[2]

        # continue : X 허용 범위를 벗어난 경우
        if new_x < 0 or new_x >= x_size:
            continue

        # continue : Y 허용 범위를 벗어난 경우
        if new_y < 0 or new_y >= y_size:
            continue

        # continue : Z 허용 범위를 벗어난 경우
        if new_z < 0 or new_z >= z_size:
            continue

        # continue : 이미 방문한 칸인 경우
        if visited[new_z][new_y][new_x]:
            continue

        # continue : 빈 칸을 방문한 경우
        if tomatoes[new_z][new_y][new_x] == -1:
            continue

        queue.append([new_z, new_y, new_x, cur_day + 1])
        visited[new_z][new_y][new_x] = True
        colored_size += 1
        target_day = cur_day + 1

if colored_size + empty_size == total_size:
    if colored_size:
        print(target_day)
    else:
        print(0)
else:
    print(-1)

File: 골드 5 (7문제)/7569_토마토.c
// 설명 링크 : https://moon-gd.tistory.com/80
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#define MAX 101

typedef struct node {
	int x;
	int y;
	int z;
	int day;
}node;
node queue[MAX * MAX * MAX];

// 큐 관련 함수
int front = 0, rear = 0;
void enque(int, int, int, int);
node deque();

int tomatoes[MAX][MAX][MAX];  // 토마토 상자 배열
int visited[MAX][MAX][MAX];  // 방문 배열

// 이동 관련 방향 배열
int dir_x[6] = { 0, 0, 0, 1, 0, -1 };
int dir_y[6] = { 0, 0, -1, 0, 1, 0 };
int dir_z[6] = { 1, -1, 0, 0, 0, 0 };

int main() {
	// 반복문 순회를 위한 임시 전역 변수
	int i, j, k;

	// 토마토 상자 크기 확인
	int x_size, y_size, z_size;
	scanf("%d %d %d", &x_size, &y_size, &z_size);

	// 공간 정보 체크를 위한 변수
	int total_size, colored_size = 0, empty_size = 0;
	total_size = x_size * y_size * z_size;

	// 상자 정보 입력 받기
	for (k = 0; k < z_size; k++) {
		for (j = 0; j < y_size; j++) {
			for (i = 0; i < x_size; i++) {
				scanf("%d", &tomatoes[i][j][k]);

				// 토마토가 있는 칸인 경우
				if (tomatoes[i][j][k] == 1) {
					// 큐에 넣어주고 방문 정보 체크
					enque(i, j, k, 0);
					visited[i][j][k] = 1;
				}
				// 빈 칸인 경우
				else if (tomatoes[i][j][k] == -1) {
					// 빈 영역 ++
					empty_size++;
				}
			}
		}
	}

	// 색칠 영역 ++
	colored_size = rear;

	while (front != rear) {
		// 큐에서 익은 토마토 정보 하나 빼기
		node temp = deque();

		int cur_x = temp.x;
		int cur_y = temp.y;
		int cur_z = temp.z;
		int cur_day = temp.day;

		for (i = 0; i < 6; i++) {
			int new_x = cur_x + dir_x[i];
			int new_y = cur_y + dir_y[i];
			int new_z = cur_z + dir_z[i];

			// X 범위 유효성 검사
			if (new_x < 0 || new_x >= x_size) {
				continue;
			}

			// Y 범위 유효성 검사
			if (new_y < 0 || new_y >= y_size) {
				continue;
			}

			// Z 범위 유효성 검사
			if (new_z < 0 || new_z >= z_size) {
				continue;
			}

			// 이미 방문한 곳인 경우
			if (visited[new_x][new_y][new_z]) {
				continue;
			}

			// 빈 칸인 경우
			if (tomatoes[new_x][new_y][new_z] == -1) {
				continue;
			}

			// 위의 반례에 모두 해당하지 않는 경우
			// 즉, 익지 않은 새로운 토마토를 찾았을 경우
			enque(new_x, new_y, new_z, cur_day + 1);

			// 색칠 영역 추가 + 방문 정보 체크
			colored_size++;
			visited[new_x][new_y][new_z] = 1;
		}
	}

	if (colored_size + empty_size == total_size) {
		if (colored_size) {
			printf("%d", queue[rear - 1].day);
		}
		else {
			printf("0");
		}
	}
	else {
		printf("-1");
	}

	return 0;
}

void enque(int x, int y, int z, int day) {
	node temp = { x, y, z, day };
	queue[rear++] = temp;
}

node deque() {
	return queue[front++];
}